File: static/gallery_res/laplacian/laplacian.py
"""
L_i,j = { w_i,j if (i,j) ∈ E

L_i,i = -sum_(l for l != i) L_i,l

where

L ∈ ℝ^(n×n)
w ∈ ℝ^(n×n): edge weight matrix
E ∈ {ℤ²} index: edges
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


def laplacian(w, E):
    """
    :param :w : edge weight matrix
    :param :E : edges
    """
    w = np.asarray(w, dtype=np.float64)

    n = w.shape[1]
    assert w.shape == (n, n)
    assert isinstance(E, list) and len(E) > 0
    assert len(E[0]) == 2

    _Lij_0 = []
    _Lvals_0 = []
    for i in range(1, n+1):
        for j in range(1, n+1):
            if (i-1, j-1) in E:
                _Lij_0.append((i-1, j-1))
                _Lvals_0.append(w[i-1, j-1])
    _sparse_0 = scipy.sparse.coo_matrix((_Lvals_0, np.asarray(_Lij_0).T), shape=(n, n))
    L = _sparse_0

    for i in range(1, n+1):
        _sum_0 = 0
        for l in range(1, L.shape[0]+1):
            if(l != i):
                _sum_0 += L.tocsr()[i-1, l-1]
        _Lij_0.append((i - 1, i - 1))
        _Lvals_0.append(-_sum_0)
    L = scipy.sparse.coo_matrix((_Lvals_0, np.asarray(_Lij_0).T), shape=(n, n))


    return L


def generateRandomData():
    n = np.random.randint(10)
    w = np.random.randn(n, n)
    E = []
    E_0 = np.random.randint(1, 10)
    for i in range(E_0):
        E.append((np.random.randint(10), np.random.randint(10)))
    return w, E


if __name__ == '__main__':
    w, E = generateRandomData()
    print("w:", w)
    print("E:", E)
    func_value = laplacian(w, E)
    print("func_value: ", func_value)

File: static/gallery_res/laplacian/laplacian.cpp
/*
L_i,j = { w_i,j if (i,j) ∈ E

L_i,i = -sum_(l for l != i) L_i,l

where

L ∈ ℝ^(n×n)
w ∈ ℝ^(n×n): edge weight matrix
E ∈ {ℤ²} index: edges
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

/**
 * laplacian
 *
 * @param w  edge weight matrix
 * @param E  edges
 * @return L
 */
Eigen::SparseMatrix<double> laplacian(
    const Eigen::MatrixXd & w,
    const std::set<std::tuple< int, int > > & E)
{
    const long n = w.cols();
    assert( w.rows() == n );
    assert( w.cols() == n );

    Eigen::SparseMatrix<double> L(n, n);
    std::vector<Eigen::Triplet<double> > tripletList_L;
    for( int i=1; i<=n; i++){
        for( int j=1; j<=n; j++){
            if(E.find(std::tuple< int, int >(i-1, j-1)) != E.end()){
                tripletList_L.push_back(Eigen::Triplet<double>(i-1, j-1, w(i-1, j-1)));
            }
        }
    }
    L.setFromTriplets(tripletList_L.begin(), tripletList_L.end());


    for( int i=1; i<=n; i++){
        double _sum_0 = 0;
        for(int l=1; l<=L.cols(); l++){
            if(l != i){
                _sum_0 += L.coeff(i-1, l-1);
            }
        }
        tripletList_L.push_back(Eigen::Triplet<double>(i-1, i-1, -_sum_0));
    }
    L.setFromTriplets(tripletList_L.begin(), tripletList_L.end());


    return L;
}


void generateRandomData(Eigen::MatrixXd & w,
    std::set<std::tuple< int, int > > & E)
{
    const int n = rand()%10;
    w = Eigen::MatrixXd::Random(n, n);
    const int E_0 = rand()%10;
    for(int i=0; i<E_0; i++){
        E.insert(std::make_tuple(rand()%10, rand()%10));
    }
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    Eigen::MatrixXd w;
    std::set<std::tuple< int, int > > E;
    generateRandomData(w, E);
    Eigen::SparseMatrix<double> func_value = laplacian(w, E);
    std::cout<<"func_value:\n"<<func_value<<std::endl;
    return 0;
}